design/rest-api-guideline.md

```diff
@@ -0,0 +1,389 @@
+|Verb|Description|Idempotent|
+|---|---|---|
+|GET|Requests the current state of a resource|Yes|
+|POST|Creates a new resource or submit a command|No|
+|PUT|Replaces a resource state|Yes|
+|DELETE|Removes a resource|Yes|
+
+[TODO]: Shall we encourage PATCH for partial updates. That could make things easier for clients and avoid confusion when using PUT to update resources, where PUT
+should really be about replacing a resource, usually after a GET request to obtain the current resource representation.
+
+#### GET
+
+HTTP `GET` should be used to obtain current state of a resource. It is usually related with operations on resources that return either a list or a single representation, or start some action that does not change the resource state.
+
+##### Expected Status Codes
+
+It is expected that APIs responding to HTTP `GET` requests respect the following HTTP status codes to communicate the result of the operation:
+
+|Status Codes|
+|---|
+|200|
+|204|
+|307|
+|400|
+|401|
+|403|
+|404|
+|500|
+
+##### Notes on `201` Status Code and Actions using `GET`
+
+`GET` can represent some action supported by a resource (see [Controller Resource](#controllerresource) for more details) and eventually
+create resources as a consequence of the action. In this case, developers can use `201` on `GET` responses to communicate to clients the result of the operation.
+
+Under this circumstance, the developer should consider the same principles for `POST` requests especially in regards to managing
+conflicts.
+
+##### Notes on `204` Status Code and Actions using `GET`
+
+`GET` can represent some action supported by a resource (see [Controller Resource](#controllerresource) for more details) and eventually start
+some processing asynchronously without necessarily return a response body. In this case, developers can use `204` on `GET` responses to communicate to clients the result of the operation.
+
+#### POST
+
+HTTP `POST` should be used to create new resources. It is a non-safe and non-idempotent operation and clients should be aware of that.
+
+Successful `POST` requests should always result in a `201` status code and empty response. The response should also contain a reference
+to the resource that was created:
+
+```
+HTTP/1.1 201 Created
+Content-Length: 0
+Location: http://mykeycloak.com/apis/users/{version}/1234
+```
+
+##### Expected Status Codes
+
+It is expected that APIs responding to HTTP `POST` requests respect the following HTTP status codes to communicate the result of the operation:
+
+|Status Codes|
+|---|
+|201|
+|400|
+|401|
+|403|
+|404|
+|409|
+|500|
+
+#### PUT
+
+HTTP `PUT` should be used to replace a resource state.
+
+Successful `PUT` requests should always result in a `204` status code and empty response.
+
+```
+HTTP/1.1 204 No Content
+Content-Length: 0
+```
+
+##### Expected Status Codes
+
+It is expected that APIs responding to HTTP `POST` requests respect the following HTTP status codes to communicate the result of the operation:
```

Comment:
"... that APIs responding to HTTP ***`PUT`*** ..."

Reply:
Marking as resolved, because it was already updated.